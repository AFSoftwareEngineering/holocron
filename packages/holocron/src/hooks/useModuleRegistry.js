/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';

import {
  REGISTRY_MODULES_KEY, REGISTRY_MODULE_MAP_KEY, REGISTRY_MODULE_BLOCKED_KEY, MODULES_STORE_KEY,
} from '../ducks/constants';
import reducer, {
  getModule,
  getModules,
  getModuleMap,
  setModuleMap,
  registerModule,
  blockModule,
  getBlockedModules,
  isModuleBlocked,
  resetModuleRegistry,
  createInitialState,
} from '../ducks/registry';
import {
  getModuleMap as hydrateModuleMap,
  getModules as hydrateModules,
  getModuleBlockList as hydrateBlockedModules,
} from '../moduleRegistry';

export function createRegistryActions(state, dispatch) {
  return {
    getModuleMetaData(moduleName) {
      return this.getModuleMap().getIn([MODULES_STORE_KEY, moduleName]);
    },
    getModule(moduleName) {
      return getModule(moduleName)(state);
    },
    getModules() {
      return getModules()(state);
    },
    getModuleMap() {
      return getModuleMap()(state);
    },
    getBlockedModules() {
      return getBlockedModules()(state);
    },
    isModuleBlocked(moduleUrl) {
      return isModuleBlocked(moduleUrl)(state);
    },
    blockModule(moduleUrl) {
      return dispatch(blockModule(moduleUrl));
    },
    registerModule(moduleName, module) {
      return dispatch(registerModule(moduleName, module));
    },
    setModuleMap(newModuleMap) {
      return dispatch(setModuleMap(newModuleMap));
    },
    resetModuleRegistry(newModules, newModuleMap) {
      return dispatch(resetModuleRegistry(newModules, newModuleMap));
    },
  };
}

export default function useModuleRegistry(
  holocronModuleMap = hydrateModuleMap(),
  holocronModules = hydrateModules(),
  blockedModules = hydrateBlockedModules()
) {
  const [state, dispatch] = React.useReducer(
    reducer,
    createInitialState(holocronModuleMap, holocronModules, blockedModules)
  );
  return React.useMemo(() => ({
    ...createRegistryActions(state, dispatch),
    dispatch,
    modules: state.get(REGISTRY_MODULES_KEY),
    moduleMap: state.get(REGISTRY_MODULE_MAP_KEY),
    blocked: state.get(REGISTRY_MODULE_BLOCKED_KEY),
  }), [state, dispatch]);
}
