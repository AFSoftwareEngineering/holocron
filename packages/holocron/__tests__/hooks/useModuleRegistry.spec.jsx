/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import { renderHook, act } from '@testing-library/react-hooks';
import useModuleRegistry from '../../src/hooks/useModuleRegistry';

describe('useModuleRegistry', () => {
  function Module() { }

  test('returns state', () => {
    const { result } = renderHook(() => useModuleRegistry());
    expect(result.current).toMatchSnapshot();
  });

  test('adds a holocron module and removes it', () => {
    const moduleName = 'my-module';
    const { result } = renderHook(() => useModuleRegistry());
    expect(result.current.getModule(moduleName)).toBeUndefined();
    act(() => {
      result.current.registerModule(moduleName, Module);
    });
    expect(result.current.getModule(moduleName)).toEqual(expect.any(Function));
    act(() => {
      result.current.resetModuleRegistry();
    });
    expect(result.current.getModule(moduleName)).toBeUndefined();
  });
});
