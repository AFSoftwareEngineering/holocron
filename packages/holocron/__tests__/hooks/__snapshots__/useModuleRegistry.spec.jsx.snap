// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useModuleRegistry adds a holocron module and removes it 1`] = `
Object {
  "blockModule": [Function],
  "blocked": Immutable.List [],
  "getBlockedModules": [Function],
  "getModule": [Function],
  "getModuleMap": [Function],
  "getModules": [Function],
  "isModuleBlocked": [Function],
  "moduleMap": Immutable.Map {
    "modules": Immutable.Map {},
  },
  "modules": Immutable.Map {
    "my-module": [Function],
  },
  "registerModule": [Function],
  "resetModuleRegistry": [Function],
  "setModuleMap": [Function],
}
`;

exports[`useModuleRegistry default state 1`] = `
Object {
  "blockModule": [Function],
  "blocked": Immutable.List [],
  "getBlockedModules": [Function],
  "getModule": [Function],
  "getModuleMap": [Function],
  "getModules": [Function],
  "isModuleBlocked": [Function],
  "moduleMap": Immutable.Map {
    "modules": Immutable.Map {},
  },
  "modules": Immutable.Map {},
  "registerModule": [Function],
  "resetModuleRegistry": [Function],
  "setModuleMap": [Function],
}
`;
